module4-authentication-and-security/r1.2-adding-authorization-layer/README.md

```diff
@@ -0,0 +1,222 @@
+
+- Authentication is used by a server when the server needs to know exactly who is accessing their information or site.
+- Authentication is used by a client when the client needs to know that the server is the system it claims to be.
+- In authentication, the user or computer has to prove its identity to the server or client.
+- Usually, authentication by a server entails the use of a user name and password. Other ways to authenticate can be through cards, retina scans, voice recognition, and fingerprints.
+- Authentication by a client usually involves the server giving a certificate to the client in which a trusted third party such as Verisign or Thawte states that the server belongs to the entity (such as a bank) that the client expects it to. This is required to initiate secure communication (HTTPs) between client and server.
+- Authentication does not determine what tasks the individual can do or what files the individual can see (aka scopes). Authentication merely identifies and verifies who the person or system is.
+
+## **Role-based authorization**
+
+By role-based authorization, we assign a role or roles to the user either when we register them, or by some other mechanism that happens later to elevate a user access level.
+
+In simple terms, we assign an attribute to the User model, that would tell us later when he/she signs in, what is their access level, or role. Based on that, our app can determine whether they can access that resource/endpoint.
+
+Let's assume we have a vlog web app. We have 3 access levels:
+
+- An anonymous user can watch videos and search on the website without signing up.
+- To upload videos or flag content, the user needs to create an account.
+- Our app needs moderators as well, who are users that can delete videos in case they violate the usage agreement, or are inappropriate.
+- And finally, our app needs admins, who have the ability to make moderators by elevating other users to a moderator access level.
+
+From these requirements, we can distinguish our app structure as follows:
+
+- Some endpoints wouldn't have any authentication guard, any request can reach them. Which are endpoints to read a list of videos, watch a specific video, or search for videos.
+- Certain endpoints like `/upload` and `/flag` will need an authentication guard. This is to check if we have a JWT in the request or a `user` in the session. If there isn't, they shouldn't be able to reach that resource.
+- Endpoints like `/delete` would need an authorization guard. Which is a guard that contains the authentication guard, plus it also checks if the user is a `moderator`, before letting them complete the action. Otherwise, it shouldn't let them through.
+- To elevate a user, an endpoint like `/elevate` would have the same authorization guard, but instead it will check if the user is an `admin`.
+
+### **Implementation**
+
+To implement this as simply as possible, we need to first define our roles as numbers. That is mainly to:
```

Comment:
i don't think you should ever do this. coding any type of enum as meaningless numbers is exponentially worse than line 49 (hardcoded strings). 50 is also not true. 51 is also not true.

please don't do this -- it's definitely encouraging the wrong practices. a lot of things can go wrong when you use numbers, including facilitating bugs caused by typing the wrong number, making any console.log/data inspection a lot more unreadable, etc.

Reply:
sql dbs have enums for this

if you're using mongo you can use schema validation to enforce enums from what i've seen

https://stackoverflow.com/questions/28393582/store-enum-mongodb

Reply:
https://docs.mongodb.com/manual/core/schema-validation/

Reply:
I am trying to understand your disagreement here, if it is about having roles identified by integers, then I have to disagree. Many of the roll-based access control implementations use a combination of:
- Roles table
- Permissions table
- Users Table

And they are related by foreign key constraints: User.has(role) Role.has(permission)

![image](https://user-images.githubusercontent.com/12988551/135766430-511cb74e-fa9e-42b1-baca-ef8031fd5af6.png)
![image](https://user-images.githubusercontent.com/12988551/135766435-c6d66387-5021-4029-94b9-6d2a30be3545.png)

In no case, one should type a number in a hardcoded approach, and one can mess up writing a number, they would more likely mess writing admin if that is the case. It is usually by querying the database for roles and rendering them in an options element with the role_id.

If there is any resource that strictly points this out as a wrong practice, please share.

If it is about having the roles as a code-based enum, then yes it is pointed out below that this isn't the best practice, and they need to be in a database. But I tend to simplify the implementation to make it understandable. However, I'm happy to change this to a database model with a more solid example.